framework/src/executioners/Eigenvalue.C

```diff
@@ -114,31 +224,160 @@ Eigenvalue::execute()
     _eigen_problem.petscOptionsInserted() = true;
   }
 #endif
+}
+
+void
+Eigenvalue::setFreeNonlinearPowerIterations(unsigned int free_power_iterations)
+{
+#if !PETSC_RELEASE_LESS_THAN(3, 12, 0)
+  // Master app has the default data base
+  if (!_app.isUltimateMaster())
+    PetscOptionsPush(_eigen_problem.petscOptionsDatabase());
 #endif
 
-  Steady::execute();
+  Moose::SlepcSupport::setFreeNonlinearPowerIterations(free_power_iterations);
+
+#if !PETSC_RELEASE_LESS_THAN(3, 12, 0)
+  if (!_app.isUltimateMaster())
+    PetscOptionsPop();
+#endif
 }
 
 void
-Eigenvalue::postSolve()
+Eigenvalue::clearFreeNonlinearPowerIterations()
 {
-#ifdef LIBMESH_HAVE_SLEPC
-  if (_normalization)
+#if !PETSC_RELEASE_LESS_THAN(3, 12, 0)
+  // Master app has the default data base
+  if (!_app.isUltimateMaster())
+    PetscOptionsPush(_eigen_problem.petscOptionsDatabase());
+#endif
+
+  Moose::SlepcSupport::clearFreeNonlinearPowerIterations(_pars);
+
+#if !PETSC_RELEASE_LESS_THAN(3, 12, 0)
+  if (!_app.isUltimateMaster())
+    PetscOptionsPop();
+#endif
+}
+
+void
+Eigenvalue::checkIntegrity()
+{
+  // check to make sure that we don't have any time kernels in eigenvaue simulation
+  if (_eigen_problem.getNonlinearSystemBase().containsTimeKernel())
+    mooseError("You have specified time kernels in your eigenvaue simulation");
+}
+
+void
+Eigenvalue::outputInverseEigenvalue(bool inverse)
+{
+  _eigen_problem.outputInverseEigenvalue(inverse);
+}
+
+#endif
+
+void
+Eigenvalue::execute()
+{
+#if LIBMESH_HAVE_SLEPC
+  // Recovering makes sense for only transient simulations since the solution from
+  // the previous time steps is required.
+  if (_app.isRecovering())
+    return;
+
+  preExecute();
+
+  // The following code of this function is copied from "Steady"
+  // "Eigenvalue" implementation can be considered a one-time-step simulation to
+  // have the code compatible with the rest moose world.
+  _eigen_problem.advanceState();
+
+  // First step in any eigenvalue state solve is always 1 (preserving backwards compatibility)
+  _time_step = 1;
+
+#ifdef LIBMESH_ENABLE_AMR
+
+  // Define the refinement loop
+  unsigned int steps = _eigen_problem.adaptivity().getSteps();
```

Comment:
Could use `auto` here in case you ever wanted to change the return type of `getSteps`. But I guess you will probably never do more than 2^32 steps :laughing: